File: Competitive Programming 3/Section 1.4/Palidromes/10945 - Mother bear/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

set<char> puntuaction = {
    '.',
    ',',
    '!',
    '?',
    ' '
};

bool palindrome(const string &s){
    for(int i = 0 , j = s.size()-1 ; i < j ; i++,j--){
        //ignore puntuactions
        for(;i<j&&puntuaction.find(s[i])!=puntuaction.end();i++);
        for(;i<j&&puntuaction.find(s[j])!=puntuaction.end();j--);
        //check if equivalent (ignoring casing)
        if(i<j&&tolower(s[i])!=tolower(s[j]))
            return false;
    }
    return true;
}

int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    string sentence;
    while(getline(cin,sentence),sentence!="DONE"){
        cout << (palindrome(sentence)?"You won't be eaten!":"Uh oh..") << "\n";
    }
    return 0;
}

File: Competitive Programming 3/Section 1.4/Palidromes/10945 - Mother bear/Python/main.py
puntuaction = {'.',',','!','?',' '}
def palindrome(s):
    i,j=0,len(s)-1
    while i < j:
        #ignore puntuactions
        while i<j and s[i] in puntuaction:
            i+=1
        while i<j and s[j] in puntuaction:
            j-=1
        #check if equivalent (ignoring casing)
        if i<j and s[i].lower()!=s[j].lower():
            return False
        i+=1
        j-=1
    return True
def main():
    while True:
        s = input()
        if s == "DONE":
            break
        print("You won't be eaten!" if palindrome(s) else "Uh oh..")
main()